aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -570,6 +574,133 @@ struct cpu_scatter_gather_base_kernel {
   }
 };
 
+// Note [scatter reduce optimization]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+//
+// 1. initiative: optimize scatter_reduce optimization on PyG
+//   `scatter_add` is extensively used on 'message passing' when
+//   aggregating info. The `index` tensor is extended which means
+//   the aggregation is on rowwise.
+//
+// 2. implementation: map `scatter_reduce` to `spmm` reduce
```

Comment:
The key factor to choose spmm or vanilla implementation is the size of the inner dimension, which is `K` in the code, which means the channel size.

If channel size is not big enough, sorting it would not be beneficial.

I put a threshold of `16` in the kernel which is a conservative number (the tested threshold is 3 or 4 on my machine).